### PR TITLE
TRUNK-5208: Replaced older version to 1.8.4-SNAPSHOT and the newer ve…

### PR DESCRIPTION
**Pull request by ODORA0**
…rsion to 1.8.4. https://issues.openmrs.org/browse/TRUNK-5208

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-

## Checklist: I completed these to help reviewers :)

- [ ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/test/java/org/openmrs/module/ModuleUtilIT.java [hunk 1]
```diff
@@ -372,8 +372,8 @@ public void compareVersion_shouldCorrectlyComparingTwoVersionNumbers() throws Ex
 	 */
 	@Test
 	public void compareVersion_shouldTreatSNAPSHOTAsEarliestVersion() throws Exception {
-		String olderVersion = "1.8.3";
-		String newerVersion = "1.8.4-SNAPSHOT";
+		String olderVersion = "1.8.4~SNAPSHOT";
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/ModuleUtilIT.java]: Did you intend to remove the `-`?

Comment (ODORA0) [reply api/src/test/java/org/openmrs/module/ModuleUtilIT.java]: Yes, that's my bad I guess.
Should i replace it with -?

Comment (samuelmale) [reply api/src/test/java/org/openmrs/module/ModuleUtilIT.java]: Did you have any reason for doing so?

Comment (ODORA0) [reply api/src/test/java/org/openmrs/module/ModuleUtilIT.java]: I actually thought the tilde has in most cases the same definition a the hyphen in java

Comment (samuelmale) [reply api/src/test/java/org/openmrs/module/ModuleUtilIT.java]: And are you sure the SNAPSHOT version is evaluated older Version?

Comment (ODORA0) [reply api/src/test/java/org/openmrs/module/ModuleUtilIT.java]: Yes plain sure as ticket description stated it that way.

Comment (samuelmale) [reply api/src/test/java/org/openmrs/module/ModuleUtilIT.java]: Well I have not had time to look at the ticket description but normally SNAPSHOT is the earliest. What do you understand with this as part of this testcase name `shouldTreatSNAPSHOTAsEarliestVersion`?

Comment (ODORA0) [reply api/src/test/java/org/openmrs/module/ModuleUtilIT.java]: Have you gone through the ticket description to know why SNAPSHOT is made older version? :)

**Comment by ODORA0** [comment]
@mozzy11 could you review? I don't understand the call for the CI build failure, unless i did not make the changes rightly.

**Comment by mozzy11** [comment]
Yes , now that the expected result from that test Unit. Now got into the actual method and make changes so that that test passes. Do not make any more changes to the test , just modify the method until the test unit passes

**Comment by mozzy11** [comment]
this is the method your suppsed towork on now  `ModuleUtil.compareVersion(newerVersion, olderVersion)`

**Comment by ODORA0** [comment]
@mozzy11 I am not able to locate the given method so as to implement the change. All I see is the testcase name with reference to the class name. Could you help me locate it's path?

**Comment by mozzy11** [comment]
seroiusly?? how can you fail to locate a method.
https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/module/ModuleUtil.java

**Comment by ODORA0** [comment]
@mozzy11 Am having trouble making the required changes in the method, could you please guide.

**Comment by mozzy11** [comment]
@ODORA0 

take deeper look at the code in this method. 
```

public static int compareVersion(String version, String value) {
		try {
			if (version == null || value == null) {
				return 0;
			}
			
			List<String> versions = new ArrayList<>();
			List<String> values = new ArrayList<>();
			String separator = "-";
			
			// strip off any qualifier e.g. "-SNAPSHOT"
			int qualifierIndex = version.indexOf(separator);
			if (qualifierIndex != -1) {
				version = version.substring(0, qualifierIndex);
			}
			
			qualifierIndex = value.indexOf(separator);
			if (qualifierIndex != -1) {
				value = value.substring(0, qualifierIndex);
			}
			
			Collections.addAll(versions, version.split("\\."));
			Collections.addAll(values, value.split("\\."));
			
			// match the sizes of the lists
			while (versions.size() < values.size()) {
				versions.add("0");
			}
			while (values.size() < versions.size()) {
				values.add("0");
			}
			
			for (int x = 0; x < versions.size(); x++) {
				String verNum = versions.get(x).trim();
				String valNum = values.get(x).trim();
				Long ver = NumberUtils.toLong(verNum, 0);
				Long val = NumberUtils.toLong(valNum, 0);
				
				int ret = ver.compareTo(val);
				if (ret != 0) {
					return ret;
				}
			}
		}
		catch (NumberFormatException e) {
			log.error("Error while converting a version/value to an integer: " + version + "/" + value, e);
		}
		
		// default return value if an error occurs or elements are equal
		return 0;
	}
```
make sure the the logic will take a` SnapshotVersion ` as older. just try to study that logic and make small change

File: api/src/main/java/org/openmrs/module/ModuleUtil.java [hunk 2]
```diff
@@ -416,7 +416,7 @@ public static void checkRequiredVersion(String version, String versionRange) thr
 	 */
 	public static int compareVersion(String version, String value) {
 		try {
-			if (version == null || value == null) {
+			if (version == "1.8.4-SNAPSHOT" || value == "1.8.4") {
```
Comment (mozzy11) [inline api/src/main/java/org/openmrs/module/ModuleUtil.java]: @ODORA0 what were you trying to do here??

File: api/src/main/java/org/openmrs/module/ModuleUtil.java [hunk 3]
```diff
@@ -440,10 +440,10 @@ public static int compareVersion(String version, String value) {
 			
 			// match the sizes of the lists
 			while (versions.size() < values.size()) {
-				versions.add("0");
+				versions.add("1.8.4-SNAPSHOT");
```
Comment (mozzy11) [inline api/src/main/java/org/openmrs/module/ModuleUtil.java]: what were you trying to do here??

Comment (ODORA0) [reply api/src/main/java/org/openmrs/module/ModuleUtil.java]: Thought that would set snapshot older by changing that line

File: api/src/main/java/org/openmrs/module/ModuleUtil.java (hunk 2 above)
Comment (ODORA0) [reply api/src/main/java/org/openmrs/module/ModuleUtil.java]: I believe this change isn't necessary as well, guess discarding the change would be good @mozzy11

Comment (mozzy11) [reply api/src/main/java/org/openmrs/module/ModuleUtil.java]: ya is unnecesary

**Comment by mozzy11** [comment]
@ODORA0 , first undo your changes in that method and try switchig the positions of the variables 
**ver** and **val** in this statement
`int ret = ver.compareTo(val);`
and run the test and see if it passes.(line 455)

**Comment by ODORA0** [comment]
@mozzy11 I changed the above line to `int ret = val.compareTo(ver);` though the tests fail and results into this  [pastebin](https://pastebin.com/tMncEGYx) failure

**Comment by ibacher** [comment]
Travis tests are failing for this. While this bug itself should be fixed, I'm closing this PR for now.

**Pull request closed without merging**